usb: fix bulk endpoint configuration for high-speed capable device

jfischer-no opened the pull request:
In the current USB device support, the sizes of bulk endpoint
are mostly configure through Kconfig and do not care if a device
is high-speed capable. The information if a USB device controller
supports high-speed comes from devicetree. Add a Kconfig option to
map this information and configure bulk endpoint sizes
accordingly.

jfischer-no committed:
scripts/Kconfig: add dt_node_str_prop_equals Kconfig function

Add function to check if a property of a node
is equal to a passed string.

The motivation is that in current USB device support
the sizes of bulk endpoints are configured in Kconfig.
The information if a USB device controller supports
high-speed comes from devicetree. This information must be
mapped in controller driver Kconfig and corresponding options
in USB device stack configured must be adjusted.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: fix bulk endpoint configuration for high-speed capable device

In the current USB device support, the sizes of bulk endpoint
are mostly configure through Kconfig and do not care if a device
is high-speed capable. The information if a USB device controller
supports high-speed comes from devicetree. Add a Kconfig option to
map this information and configure bulk endpoint sizes
accordingly.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
dts: same70: default USB HS controller to high-speed

USB HS controller is high-speed capable, default it to high-speed.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

aurel32 reviewed (changes requested):
Note that the size of the endpoint should be dynamic as a high-speed device can be connected as low-speed. Anyway that's a step in the right direction.

However for that reason we should not default the SAME70 controller to high speed in the DT, the device won't work properly if it is connected as low speed (see #11000 for more details). Leaving low speed as the default means the device will always work, albeit with lower throughput.

mbolivar-nordic commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -388,6 +388,32 @@ def dt_node_int_prop(kconf, name, path, prop, unit=None):
     if name == "dt_node_int_prop_hex":
         return hex(_node_int_prop(node, prop, unit))
 
+def dt_node_prop_equal_str(kconf, _, path, prop, val):
```
I think for consistency with dt_node_int_prop_foo, this should be called dt_node_str_prop_equals or something like that.

jfischer-no replied:
Changed to dt_node_str_prop_equals.

mbolivar-nordic reviewed (commented):
Looks good, just one bikeshed on naming the new kconfig function.

jfischer-no commented:
> Note that the size of the endpoint should be dynamic as a high-speed device can be connected as low-speed. Anyway that's a step in the right direction.
>
> However for that reason we should not default the SAME70 controller to high speed in the DT, the device won't work properly if it is connected as low speed (see #11000 for more details). Leaving low speed as the default means the device will always work, albeit with lower throughput.

We do not support low-speed, you probably mean only full-speed capable host controller, but these days it is unlikely that anyone will have such a problem. But I will change option USB_DC_HAS_HS_SUPPORT so that one can disable it from menuconfig. Current device support implementation will remain such that either high-speed or full-speed are supported, not both at the same time.

jfischer-no committed:
drivers: usb_dc_sam_usbhs: default to full-speed

Use DT_INST_ENUM_IDX_OR and always default to full-speed
if CONFIG_USB_DC_HAS_HS_SUPPORT is not set or maximum-speed
property is not defined.
Remove low-speed setting since device stack does not support it.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

mbolivar-nordic reviewed (approved):
For the DT portions

aurel32 commented:
> > Note that the size of the endpoint should be dynamic as a high-speed device can be connected as low-speed. Anyway that's a step in the right direction.
> > However for that reason we should not default the SAME70 controller to high speed in the DT, the device won't work properly if it is connected as low speed (see #11000 for more details). Leaving low speed as the default means the device will always work, albeit with lower throughput.
> 
> We do not support low-speed, you probably mean only full-speed capable host controller, but these days it is unlikely that anyone will have such a problem. But I will change option USB_DC_HAS_HS_SUPPORT so that one can disable it from menuconfig. Current device support implementation will remain such that either high-speed or full-speed are supported, not both at the same time.

Sorry, I meant-full speed vs high-speed indeed. My point is that high-speed support is far from ready in Zephyr, and we should not enable it by default until it has been fully validated. Your patch series is a step in the right direction, but still are still missing, runtime speed selection (as opposed to build time) between full and high-speed mode is one thing, but there are more things like different meaning of the `bInterval` values for interrupt and isochronous transfers between full and high speed mode.

In short what I mean is that the `dts: same70: default USB HS controller to high-speed` commit should be dropped from this series. The other commits LGTM.

jfischer-no commented:
>runtime speed selection (as opposed to build time) between full and high-speed mode is one thing

This is neither necessary nor possible in current USB device stack implementation, and again, current device support implementation will remain such that either high-speed or full-speed are supported, not both at the same time.

>but there are more things like different meaning of the bInterval values for interrupt and isochronous transfers between full and high speed mode.

Units of bInterval are different, 125 µs (microframe) or 1 ms (frame), but default values in the tree are still in valid range, even if they may not be suitable.

>In short what I mean is that the dts: same70: default USB HS controller to high-speed commit should be dropped from this series. The other commits LGTM.

The commit will stay, in case of trouble fallback is possible by disabling CONFIG_USB_DC_HAS_HS_SUPPORT.
